clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1739,17 +1738,57 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     InitExprs.push_back(MemberInit.get());
   }
 
+  int getDims() {
+    int Dims = 0;
+    for (int i = MemberExprBases.size() - 1; i >= 0; --i) {
+      if (!isa<ArraySubscriptExpr>(MemberExprBases[i]))
+        break;
+      ++Dims;
+    }
+    return Dims;
+  }
+
+  int64_t getArrayIndex(int i) {
+    ArraySubscriptExpr *LastArrayRef =
+        cast<ArraySubscriptExpr>(MemberExprBases[i]);
+    Expr *LastIdx = LastArrayRef->getIdx();
+    llvm::APSInt Result;
+    SemaRef.VerifyIntegerConstantExpression(LastIdx, &Result);
+    return Result.getExtValue();
+  }
+
   void createExprForScalarElement(FieldDecl *FD) {
-    InitializedEntity ArrayEntity =
+    llvm::SmallVector<InitializedEntity, 4> InitEntities;
+
+    // For multi-dimensional arrays, an initialized entity needs to be
+    // generated for each 'dimension'. For example, the initialized entity
+    // for s.array[x][y][z] is constructed using initialized entities for
+    // s.array[x][y], s.array[x] and s.array. InitEntities is used to maintain
+    // this.
+    InitializedEntity Entity =
         InitializedEntity::InitializeMember(FD, &VarEntity);
+    InitEntities.push_back(Entity);
+
+    // Calculate dimension using ArraySubscriptExpressions in MemberExprBases.
+    // Each dimension has an ArraySubscriptExpression (maintains index)
+    // in MemberExprBases. For example, if we are currently handling element
+    // a[0][0][1], the top of stack entries are ArraySubscriptExpressions for
+    // indices 0,0 and 1, with 1 on top.
+    int Dims = getDims();
+    int NIndex = MemberExprBases.size() - 1 - (Dims - 1);
```

Comment:
These +/- 1 don't help to understand what is doing on anyway.
```suggestion
    int NIndex = MemberExprBases.size() - Dims;
```
It would be great if we manage how to name this variable so it will be easier to understand that it is actually an index in `MemberExprBases` array.